### DISCUSSION TITLE
Difference between "anon" and "authenticated" role

### DISCUSSION THREAD

**Discussion by phuoccss** [supabase/supabase#1493] [Questions] [answered]
```sql
create table private_schema.table_name

grant
    select
on private_schema.table_name to authenticated;

create view  public.view_name as
select * from private_schema.table_name;

alter view public.view_name owner to authenticated;
```

I would expect that `anon` role aka non-logged-in users to not be able to select from the public view, since I only grant select permission to `authenticated` role. However, this is not the case as postgrest still return data when users are not logged in.

So what is the difference between `anon` and `authenticated` and under what scenarios does the api client take on each role?

**Comment by steve-chavez** [comment, marked as answer]
> I would expect that anon role aka non-logged-in users to not be able to select from the public view

Can you double check that you're making the request with the `anon` role?

I can't reproduce your error. Do this:

```sql
create schema private;

create table private.entities( id int, name text);

create view  public.entities as
select * from private.entities;
```

Go to the API docs(`api/default/`), select the entities table, select Bash, select the `anon` on the "Key" dropdown. This will give you a `curl` snippet that makes the call as the `anon` role.

Then do:

```
curl 'https://<ref>.supabase.co/rest/v1/entities?select=id' \
-H "apikey: <the_JWT>" \
-H "Authorization: Bearer <the_JWT>"

{"hint":null,"details":null,"code":"42501","message":"permission denied for schema public"}  
```

Which should be the expected error from anon.

---

Sidenote:

```sql
grant
    select
on private_schema.table_name to authenticated;

create view  public.view_name as
select * from private_schema.table_name;
```

If you create a private schema, `authenticated` shouldn't have any privilege on the `private` schema tables. You only need privileges on the views of the `public` schema.

**Reply by phuoccss** [reply]
Yes, when I do the following and run the curl command, it throws error as expected.

```sql
create schema private;

create table private.table_name( id uuid);

create view  public.view_name as
select * from private.table_name;
```

I also did the following since I want to enable RLS later on for the private table, otherwise the public view will be accessed with the superuser privilege.

```sql

alter view public.view_name owner to authenticated;
```

Then, if I do not grant select permission on the `private` schema tables to `authenticated`, front-end will throw `Permission denied` error for the private table.

![image](https://user-images.githubusercontent.com/15166715/117560633-1b1d9680-b0c2-11eb-887a-f8f84c52b935.png)

Once I grant permission on the private table, it works as expected for logged-in users.

```sql
grant
    select
on private.table_name to authenticated;
```

![image](https://user-images.githubusercontent.com/15166715/117560655-54560680-b0c2-11eb-8cdb-6818b6c7f857.png)

Now the problem is even anonymous users can view the table.

![image](https://user-images.githubusercontent.com/15166715/117560699-9717de80-b0c2-11eb-966e-a66e71bf2cd6.png)

So just wondering why is this the case, since I only grant permission to `authenticated` user.

**Reply by easy-street** [reply]
by default, the `anon` role is granted all privileges on any table/view created in the `public` schema. when querying the view, access to the underlying table is determined by the privileges of the owner of the view. so the behavior you're describing is expected; more info in the postgres [Rules and Privleges docs](https://www.postgresql.org/docs/12/rules-privileges.html).

to restrict access to authenticated users only, create an RLS policy with this definition:
```
auth.role() = 'authenticated'
```

**Reply by phuoccss** [reply]
Got it. Thanks so much for the responses.

**Reply by steve-chavez** [reply]
>  when querying the view, access to the underlying table is determined by the privileges of the owner of the view. so the behavior you're describing is expected; more info in the postgres Rules and Privleges docs.

@easy-street That's true, good call.

> by default, the anon role is granted all privileges on any table/view created in the public schema.

:O But that's **not true!** Why do you state that? It can be rebutted easily, create a new table/view on the public schema and try to query it with the `anon` role, it will give `permission denied for <table|view>`. In fact, on a new project(just created one), the `anon` role doesn't even have `grant usage` on the `public` schema.

> alter view public.view_name owner to authenticated;
> Then, if I do not grant select permission on the private schema tables to authenticated, front-end will throw Permission denied error for the private table.

@universse I see what you mean. Really nor `authenticated` or `anon` should have any privilege on the `private` schema tables, since they're both meant to be API users and only use what's available on `public`.

What I'd recommend instead is to use a dedicated owner role:

```sql
CREATE ROLE public_views_owner NOINHERIT;
ALTER VIEW entities OWNER TO public_views_owner;
GRANT ALL ON private.entities to public_views_owner;
```

That way you can keep `authenticated` privileges to a minimum(see [PoLP](https://en.wikipedia.org/wiki/Principle_of_least_privilege)). It'd only need privileges on view exposed in `public`.

> Now the problem is even anonymous users can view the table.

This should only depend on whether `anon` has grant select on the table or not. `anon` should not have any privilege by default, If that's not the case, then there must be something wrong.

**Reply by easy-street** [reply]
> > by default, the anon role is granted all privileges on any table/view created in the public schema.
> 
> :O But that's **not true!** Why do you state that? It can be rebutted easily, create a new table/view on the public schema and try to query it with the `anon` role, it will give `permission denied for <table|view>`. In fact, on a new project(just created one), the `anon` role doesn't even have `grant usage` on the `public` schema.

To be clear, I was referring to _supabase_ defaults, not postgres defaults in general. To verify, I created a new supabase project just now:

- `anon` role is granted usage on the `public` schema

![image](https://user-images.githubusercontent.com/64047794/117699937-05c06f00-b18b-11eb-8123-89f001dc0c41.png)

- `anon` role is granted all privileges on new tables in the `public` schema by default

![image](https://user-images.githubusercontent.com/64047794/117700041-24266a80-b18b-11eb-8ccf-67ec7becf1c6.png)

this is consistent with the self-hosting configuration here: https://github.com/supabase/supabase/blob/master/docker/postgres/00-initial-schema.sql#L23
```
alter default privileges in schema public grant all on tables to postgres, anon, authenticated, service_role;
```

**Reply by steve-chavez** [reply]
@easy-street Reproduced :heavy_check_mark:. Many thanks for reporting that, it's a mistake indeed. I've made a PR: https://github.com/supabase/supabase/pull/1536

@universse To fix it in your current project, you can do:

```sql
alter default privileges in schema public revoke all on tables from anon;
alter default privileges in schema public revoke all on functions from anon;
alter default privileges in schema public revoke all on sequences from anon;

-- the above will only work for future database objects, to revoke current privileges 
revoke select on all tables in schema public from anon;
revoke usage on all sequences in schema public from anon;
revoke execute on all functions in schema public from anon;
```

**Reply by phuoccss** [reply]
Thanks so much for the suggestions.

**Reply by mtte** [reply]
@steve-chavez the changes you suggested and implemented in #1536 have been reverted again. Why is that? Is it still valid to revoke the privileges of the anon user or is there a reason one shouldn't do that?

**Reply by ozum** [reply]
@steve-chavez why is #1536 reverted?

**Reply by steve-chavez** [reply]
@mtte It's still valid to revoke `anon` privileges.

@ozum That was a change for the local(CLI) Supabase version, while the hosted version still used a privileged `anon`. 

I'd also like to have a non-privileged `anon` by default(this is the PostgreSQL default, and PostgREST recommendation) but nowadays doing this for Supabase would be a big breaking change. Maybe with your custom roles proposal https://github.com/supabase/supabase/issues/11745 we can have a way forward that is also backwards compatible.
